### PR TITLE
mds: catch damage to dentry's first field

### PR DESCRIPTION
**Pull request by batrick**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/mds/CDentry.cc
Comment (lxbsz) [inline src/mds/CDentry.cc:705]: Will this also be possible that the `first` equals to `last`, which are all `head` just like in https://tracker.ceph.com/issues/38452#note-10 ?

File: src/mds/CDentry.h
Comment (lxbsz) [inline src/mds/CDentry.h:163]: How about just set `bool abort = true` as default ?

File: src/mds/CDentry.cc
Comment (lxbsz) [reply src/mds/CDentry.cc:705]: Could we also check CInode's corruption in this PR ?

Comment (batrick) [reply src/mds/CDentry.cc:705]: If `first > next_snap`, it would trigger when `first == SNAP_HEAD`.

>Could we also check CInode's corruption in this PR ?

It's not a bad idea but writing dedicated tests for that kind of corruption would take time. I believe the corruption is introduced via the dentries so it's best to get a fix in for this ASAP.

**Review by kotreshhr — APPROVED** [review]

**Review by vshankar — COMMENTED** [review]
LGTM

File: src/mds/Server.cc
Comment (batrick) [reply src/mds/Server.cc:9458]: It's a good question. It's not a routine op when e.g. building a subtree. It's easier to control when the new dentry is made.

**Comment by rishabh-d-dave** [comment]
jenkins test make check

**Comment by rishabh-d-dave** [comment]
jenkins test api

**Comment by batrick** [comment]
jenkins test api

**Comment by vshankar** [comment]
jenkins test api

**Comment by batrick** [comment]
jenkins test api

**Comment by vshankar** [comment]
@batrick I haven't gone through the test run, but this test failure caught my eye which might be related: https://pulpito.ceph.com/vshankar-2023-03-03_04:39:14-fs-wip-vshankar-testing-20230303.023823-testing-default-smithi/7192234

**Comment by batrick** [comment]
> @batrick I haven't gone through the test run, but this test failure caught my eye which might be related: https://pulpito.ceph.com/vshankar-2023-03-03_04:39:14-fs-wip-vshankar-testing-20230303.023823-testing-default-smithi/7192234

This was a surprising failure I didn't see before in teuthology. We expect the mds to die but for some reason only recently see these errors AFAIK.

Anyway, I've fixed that.

I'm looking into another logical problem with this PR showing up in other test failures.

**Comment by lxbsz** [comment]
@batrick Is there ready to review ?

**Comment by batrick** [comment]
> @batrick Is there ready to review ?

No, I'm tracking down a mysterious segmentation fault in standby-replay daemons.

**Comment by batrick** [comment]
jenkins test make check arm64

**Comment by batrick** [comment]
jenkins test make check arm64

**Comment by vshankar** [comment]
Changes look good. I'll have a final rundown on the test run before merging this. Thank you, @batrick.

**Comment by batrick** [comment]
https://pulpito.ceph.com/?branch=wip-pdonnell-testing-20230323.162417

**Comment by batrick** [comment]
@vshankar config added. Going to QA now.

**Comment by batrick** [comment]
jenkins test api

File: src/mds/CDir.cc
Comment (lxbsz) [inline src/mds/CDir.cc:2565]: Is this for debug only ? The debug level is `1` ?

Comment (batrick) [reply src/mds/CDir.cc:2565]: ... yes!

Good catch!

**Comment by batrick** [comment]
https://pulpito.ceph.com/pdonnell-2023-03-29_15:15:36-fs-wip-pdonnell-testing-20230329.131031-distro-default-smithi/

without latest SQUASH commit. This is just `--suite fs --filter damage`.

@vshankar do you want more tests run?

**Comment by vshankar** [comment]
> https://pulpito.ceph.com/pdonnell-2023-03-29_15:15:36-fs-wip-pdonnell-testing-20230329.131031-distro-default-smithi/
> 
> without latest SQUASH commit. This is just `--suite fs --filter damage`.
> 
> @vshankar do you want more tests run?

Nope. Good to merge once tests pass.

**Comment by batrick** [comment]
Tests look good: https://pulpito.ceph.com/pdonnell-2023-03-29_15:15:36-fs-wip-pdonnell-testing-20230329.131031-distro-default-smithi/

The warning should be ignorelisted. I added it to `qa/tasks/cephfs: test damage to dentry's first is caught `.

Will merge when jenkins test pass.

**Comment by gregsfortytwo** [comment]
> The warning should be ignorelisted. I added it to `qa/tasks/cephfs: test damage to dentry's first is caught `.

Only for the tests we are deliberately inducing it in, right? If this turns up in the lab we definitely want to detect it!

**Comment by batrick** [comment]
> > The warning should be ignorelisted. I added it to `qa/tasks/cephfs: test damage to dentry's first is caught `.
> 
> Only for the tests we are deliberately inducing it in, right? If this turns up in the lab we definitely want to detect it!

Yes, it's induced.

**Pull request merged**
